File: Python/BinarySearch.py
def binarySearch(lista,key,inicio=0,final=None):
    if final is None:
        final = len(lista) - 1
    if inicio<=final:
        meio = (inicio+final)//2
        if lista[meio]==key:
            return meio
        if key<lista[meio]:
            return binarySearch(lista,key,inicio,meio-1)
        else:
            return binarySearch(lista,key,meio+1,final)
    return None
lista = [1,4,15,16,18,20,25,27,27,90]
n = binarySearch(lista,4)
print(n)

File: C-C++/BinarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

int binarySearch(vector<int>lista,int inicio,int final,int key){
	if (final==NULL){
		final = lista.size();
	} if (inicio<=final){
		int meio = (final + inicio)/2;
		if(lista[meio]==key){
			return meio;
		} else if(lista[meio]>key){
			cout << lista[meio] << "    " << key << endl;
			return binarySearch(lista,inicio,meio,key);
		} else{
			cout << lista[meio] << "    " << key << endl;
			return binarySearch(lista,meio,final,key);
		}
	}
	return -1;
}

int main(){
	vector<int>nums;
	nums.push_back(12);
	nums.push_back(14);
	nums.push_back(15);
	nums.push_back(19);
	nums.push_back(25);
	nums.push_back(29);
	
	cout << binarySearch(nums,0,NULL,19) << endl;
	return 0;
}

